File: csharp/xfa/01_basic_positioned_form/01_basic_positioned_form.cs
// ============================================================================
//  01_basic_positioned_form -- C# port of examples\delphi\xfa\01_basic_positioned_form
//  XFA "flavor tour" example 1 of 10: POSITIONED LAYOUT -- every subform/draw/
//  field carries layout="position" and an explicit x/y/w/h, no flow, no occur,
//  no pagination. A single-page "Employee Information" HR record with a
//  masthead, five statically placed fields (name, employee ID, department,
//  hire date, a "full-time" checkbox) bound to an <xfa:datasets> packet, and
//  a photo-placeholder box.
//
//  Mirrors the Delphi driver's exact call sequence -- the packet files were
//  already split out of the source .xdp by a one-off tool, so this driver
//  just reads the two raw XML files and hands them straight to the engine:
//
//    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
//    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
//
//  Does not rebuild LumasPdf.dll -- links against the generated P/Invoke
//  wrapper wrappers\dotnet\LumasPdf.cs (LumasPdf.Net.dll) exactly like every
//  other C# example in this project, and loads whatever LumasPdf.dll is
//  copied next to the .exe.
// ============================================================================
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class BasicPositionedForm
{
    static int RenderExample(string exeDir, string baseName, string outPdfPath)
    {
        string templatePath = Path.Combine(exeDir, baseName + ".template.xml");
        string datasetsPath = Path.Combine(exeDir, baseName + ".datasets.xml");

        Console.WriteLine("=== " + baseName + " -> " + Path.GetFileName(outPdfPath) + " ===");
        if (!File.Exists(templatePath)) { Console.WriteLine("FILE-NOT-FOUND: " + templatePath); return -100; }

        byte[] templateBuf = File.ReadAllBytes(templatePath);
        byte[] datasetsBuf = File.Exists(datasetsPath) ? File.ReadAllBytes(datasetsPath) : new byte[0];
        Console.WriteLine("template packet bytes: " + templateBuf.Length);
        Console.WriteLine("datasets packet bytes: " + datasetsBuf.Length);

        IntPtr pdf = LumasPdf.pdfNewPDF();
        if (pdf == IntPtr.Zero) { Console.WriteLine("pdfNewPDF FAILED"); return -100; }
        try
        {
            if (!LumasPdf.pdfCreateNewPDFW(pdf, outPdfPath)) { Console.WriteLine("pdfCreateNewPDFA FAILED"); return -100; }

            GCHandle hTpl = GCHandle.Alloc(templateBuf, GCHandleType.Pinned);
            int idx;
            try
            {
                idx = LumasPdf.pdfCreateXFAStreamW(pdf, "template", hTpl.AddrOfPinnedObject(), (uint)templateBuf.Length);
            }
            finally { hTpl.Free(); }
            Console.WriteLine("pdfCreateXFAStreamA(template) -> index " + idx);
            if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(template) FAILED"); return -100; }

            if (datasetsBuf.Length > 0)
            {
                GCHandle hDs = GCHandle.Alloc(datasetsBuf, GCHandleType.Pinned);
                try
                {
                    idx = LumasPdf.pdfCreateXFAStreamW(pdf, "datasets", hDs.AddrOfPinnedObject(), (uint)datasetsBuf.Length);
                }
                finally { hDs.Free(); }
                Console.WriteLine("pdfCreateXFAStreamA(datasets) -> index " + idx);
                if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(datasets) FAILED"); return -100; }
            }
            else
            {
                Console.WriteLine("(no datasets packet found -- template-only render)");
            }

            int rc = LumasPdf.pdfRenderXFAForm(pdf);
            Console.WriteLine("pdfRenderXFAForm -> " + rc + " (expected: page count >= 1)");
            if (rc < 1) { Console.WriteLine("RENDER-FAILED, code " + rc); return rc; }

            if (!LumasPdf.pdfCloseFile(pdf)) { Console.WriteLine("pdfCloseFile FAILED"); return -101; }
            Console.WriteLine("OK: wrote " + outPdfPath + " (" + rc + " page(s))");
            return rc;
        }
        finally
        {
            LumasPdf.pdfDeletePDF(pdf);
        }
    }

    static void Main()
    {
        try
        {
            string exeDir = AppDomain.CurrentDomain.BaseDirectory;
            int rc = RenderExample(exeDir, "01_basic_positioned_form", Path.Combine(exeDir, "output.pdf"));
            Console.WriteLine();
            Console.WriteLine("RESULT|01_basic_positioned_form=" + rc);
            if (rc >= 1)
            {
                Console.WriteLine("OK: rendered " + rc + " page(s). Open output.pdf and confirm the five");
                Console.WriteLine("statically-positioned fields (layout=\"position\" throughout, no flow/occur/pagination):");
                Console.WriteLine("  - Full Name    = \"Sarah J. Connor\"");
                Console.WriteLine("  - Employee ID  = \"EMP-10457\"");
                Console.WriteLine("  - Department   = \"Engineering\"");
                Console.WriteLine("  - Hire Date    = \"2021-03-15\"");
                Console.WriteLine("  - Full-time checkbox bound to FullTime=1 (checked)");
            }
            else
            {
                Console.WriteLine("FAILED, see errors above.");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine("EXCEPTION: " + e.GetType().Name + ": " + e.Message);
        }
    }
}

File: python/xfa/01_basic_positioned_form/01_basic_positioned_form.py
# ============================================================================
#  01_basic_positioned_form -- Python (ctypes) port of
#  examples\delphi\xfa\01_basic_positioned_form\01_basic_positioned_form.dpr
#
#  "Flavor tour" example 1 of 10: POSITIONED LAYOUT -- static field
#  positioning, no flow/occur/pagination. Renders a single-page "Employee
#  Information" HR record (masthead, five statically placed/data-bound
#  fields, a photo-placeholder box) through the real LumasPdf.dll:
#
#    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
#    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
#
#  The Delphi driver parses the combined .xdp and splits it into
#  <template>/<datasets> packets at run time. This Python port instead reads
#  the two packets pre-split into their own files (this session's one-off
#  split_xfa_packets tool already extracted them next to the Delphi .xdp) --
#  no XML library needed here, just two raw-byte file reads.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

HERE = os.path.dirname(os.path.abspath(__file__))


def render(template_path, datasets_path, out_pdf_path):
    template = open(template_path, "rb").read()
    datasets = open(datasets_path, "rb").read()
    print("template packet bytes:", len(template))
    print("datasets packet bytes:", len(datasets))

    pdf = L.pdfNewPDF()
    if not pdf:
        print("pdfNewPDF FAILED")
        return -100
    try:
        if L.pdfCreateNewPDFA(pdf, out_pdf_path.encode("latin-1")) == 0:
            print("pdfCreateNewPDFA FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"template", template, len(template))
        print("pdfCreateXFAStreamA(template) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(template) FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"datasets", datasets, len(datasets))
        print("pdfCreateXFAStreamA(datasets) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(datasets) FAILED")
            return -100

        rc = L.pdfRenderXFAForm(pdf)
        print("pdfRenderXFAForm ->", rc)
        if rc < 1:
            print("pdfRenderXFAForm FAILED, code", rc)
            return rc

        if L.pdfCloseFile(pdf) == 0:
            print("pdfCloseFile FAILED")
            return -101

        print("OK: wrote", out_pdf_path)
        return rc
    finally:
        L.pdfDeletePDF(pdf)


def main():
    template_path = os.path.join(HERE, "01_basic_positioned_form.template.xml")
    datasets_path = os.path.join(HERE, "01_basic_positioned_form.datasets.xml")
    out_pdf_path = os.path.join(HERE, "output.pdf")

    print("=== 01_basic_positioned_form -> output.pdf ===")
    rc = render(template_path, datasets_path, out_pdf_path)
    print("RESULT|01_basic_positioned_form=", rc)

    print()
    print("Expect a single page with these statically positioned/bound fields:")
    print('  Full Name:    "Sarah J. Connor"')
    print('  Employee ID:  "EMP-10457"')
    print('  Department:   "Engineering"')
    print('  Hire Date:    "2021-03-15"')
    print('  Full-time employee checkbox + "Employee Photo" placeholder box')


if __name__ == "__main__":
    main()
